components/tcp-sockets/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+use bindings::{
+    exports::wasi::http0_2_0::incoming_handler::Guest,
+    wasi::{
+        http0_2_0::types::{
+            Headers, IncomingRequest, OutgoingBody, OutgoingResponse, ResponseOutparam,
+        },
+        io0_2_0::poll,
+        sockets0_2_0::{
+            instance_network,
+            network::{
+                ErrorCode, IpAddressFamily, IpSocketAddress, Ipv4SocketAddress, Ipv6SocketAddress,
+            },
+            tcp_create_socket,
+        },
+    },
+};
+
+use std::net::SocketAddr;
+
+mod bindings {
+    wit_bindgen::generate!({
+            world: "http-trigger",
+            path:  "../../wit",
+    });
+    use super::Component;
+    export!(Component);
+}
+
+struct Component;
+
+impl Guest for Component {
+    fn handle(request: IncomingRequest, outparam: ResponseOutparam) {
+        // The request must have a "url" header.
+        let Some(address) = request.headers().entries().iter().find_map(|(k, v)| {
+            (k == "address")
+                .then_some(v)
+                .and_then(|v| std::str::from_utf8(v).ok())
+                .and_then(|v| v.parse().ok())
+        }) else {
+            // Otherwise, return a 400 Bad Request response.
+            return_response(outparam, 400, b"Bad Request");
+            return;
+        };
+
+        match make_request(address) {
+            Ok(()) => return_response(outparam, 200, b""),
+            Err(e) => return_response(outparam, 500, format!("{e}").as_bytes()),
+        }
+    }
+}
+
+fn make_request(address: SocketAddr) -> anyhow::Result<()> {
+    let client = tcp_create_socket::create_tcp_socket(IpAddressFamily::Ipv4)?;
+
+    client.start_connect(
+        &instance_network::instance_network(),
+        match address {
+            SocketAddr::V6(address) => {
+                let ip = address.ip().segments();
+                IpSocketAddress::Ipv6(Ipv6SocketAddress {
+                    address: (ip[0], ip[1], ip[2], ip[3], ip[4], ip[5], ip[6], ip[7]),
```

Comment:
I'm pretty sure there is a blanket `From` for smallish arrays to tuples. :shrug: